src/middleware/auth.middleware.ts

```diff
@@ -1 +1,57 @@
-// Auth middleware
+import { Request, Response } from 'express';
+import { Socket } from 'socket.io';
+import authServices from '../services/auth.services';
+
+const authMiddleware = async (
+  req : Request, 
+  res: Response, 
+  next: Function
+) => {
+  const authHeader = req.headers['Authorization'];
+
+  const token = authServices.getAuthHeader(authHeader.toString());
+  if (token === null) {
+    res.sendStatus(401);
+    return;
+  }
+  
+  try {
+    const validToken = await authServices.validateToken(token);
+    if (!validToken) {
+      res.sendStatus(401);
+      return;
+    }
+    next();
+  } catch(error) {
+    res.status(500).send({error});
+  }
+}
+
+const authSocketMiddleware = async (
+  socket: Socket, 
+  next: Function
+) => {
+  const authHeader = socket.handshake.headers['Authorization'];
+    
+  const token = authServices.getAuthHeader(authHeader.toString());
```

Comment:
buat dapetin token dari client, kayanya better pake `socket.handshake.auth` aja, nnt propertinya `socket.handshake.auth.token` (jadi di FE bisa ngeset `socket.auth = { token = "secret_token" }` terus kirim ke BE). Karena kalau pake headers dia hrs initialize duluan di FE pas bikin socket objectnya (sedangkan di FE pengennya diset tokennya setelah login). Mungkin bisa refer ke [sini](https://socket.io/docs/v3/middlewares/#sending-credentials)